### PR TITLE
Fix SA1024 + unit tests

### LINKED ISSUE #210: Diagnostic unit tests: SA1024 (ColonsMustBeSpacedCorrectly)
Implement unit tests for #30.

### LINKED ISSUE #895: SA1024 (ColonsMustBeSpacedCorrectly) doesn't report
[SA1024](http://www.stylecop.com/docs/SA1024.html) doesn't report diagnostic on the following code (there is no space before colons):
```C#
using System;

public class Foo<T>: object where T: IFormattable // 2 diagnostics should be here
{
    public Foo(): base() // one here
    {
    }
    public Foo(int x): this() // one here
    {
        Bar(value: x > 2 ? 2: 3); // one here
    }

    private int Bar(int value)
    {
    _label:
        switch (value)
        {
            case 2:
            case 3:
                return value;
            default:
                goto _label;
        }
    }
}
```

### PR DESCRIPTION
**Pull request by Maxwe11**
Closes #210
Fixes #895

### CODE REVIEW

**Comment by sharwell** [comment]
Another set of spacing tests! :smile: (There are still a whole bunch of spacing diagnostics which are "mostly" implemented but don't have tests so they are disabled.)

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1024UnitTests.cs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using StyleCop.Analyzers.SpacingRules;
+    using TestHelper;
+    using Xunit;
+
+    /// <summary>
+    /// Unit tests for <see cref="SA1024ColonsMustBeSpacedCorrectly"/>
+    /// </summary>
+
+    public class SA1024UnitTests : DiagnosticVerifier
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1024UnitTests.cs]: There's a unwanted blank line above.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1024UnitTests.cs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+            var testCode = string.Empty;
+            await this.VerifyCSharpDiagnosticAsync(testCode, EmptyDiagnosticResults, CancellationToken.None).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Verifies that the analyzer will not produce the proper diagnostics when the colons spaced correctly on same line.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> representing the asynchronous unit test.</returns>
+        [Fact]
+        public async Task TestValidSpacedColunsOnSameLineAsync()
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1024UnitTests.cs]: Typo: should be `TestValidSpacedColonsOnSameLineAsync`

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1024UnitTests.cs [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+            await this.VerifyCSharpDiagnosticAsync(testCode, EmptyDiagnosticResults, CancellationToken.None).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Verifies that the analyzer will not produce the proper diagnostics when the colons is the first or last character on the line.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> representing the asynchronous unit test.</returns>
+        [Fact]
+        public async Task TestValidSpacedColunsOnMultipleLinesAsync()
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1024UnitTests.cs]: Typo: `TestValidSpacedColonsOnMultipleLinesAsync`

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1024UnitTests.cs [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+            await this.VerifyCSharpDiagnosticAsync(testCode, EmptyDiagnosticResults, CancellationToken.None).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Verifies that the analyzer will produce the proper diagnostics when the colons not followed by space.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> representing the asynchronous unit test.</returns>
+        [Fact]
+        public async Task TestInvalidSpacedColunsMustBeFollowedAsync()
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1024UnitTests.cs]: Typo: `TestInvalidSpacedColonsMustBeFollowedAsync`

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1024UnitTests.cs [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+            await this.VerifyCSharpDiagnosticAsync(testCode, expected, CancellationToken.None).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Verifies that the analyzer will produce the proper diagnostics when the colons not preceded by space.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> representing the asynchronous unit test.</returns>
+        [Fact]
+        public async Task TestInvalidSpacedColunsMustBePrecededAsync()
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1024UnitTests.cs]: Typo: `TestInvalidSpacedColonsMustBeFollowedAsync`

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1024UnitTests.cs [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+
+            await this.VerifyCSharpDiagnosticAsync(testCode, expected, CancellationToken.None).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Verifies that the analyzer will produce the proper diagnostics when the colons must not preceded by space.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> representing the asynchronous unit test.</returns>
+        [Fact]
+        public async Task TestInvalidSpacedColunsMustNotBePrecededAsync()
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1024UnitTests.cs]: Typo: `TestInvalidSpacedColonsMustBeFollowedAsync`

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1024ColonsMustBeSpacedCorrectly.cs [hunk 7]
```diff
@@ -138,8 +138,8 @@ private void HandleColonToken(SyntaxTreeAnalysisContext context, SyntaxToken tok
                 }
             }
 
-            bool? hasPrecedingSpace = null;
-            if (!token.HasLeadingTrivia)
+            bool hasPrecedingSpace = token.HasLeadingTrivia;
+            if (!hasPrecedingSpace)
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1024ColonsMustBeSpacedCorrectly.cs]: This code will accept a comment block `/* ... */` as a space. But its not a major issue.

**Comment by vweijsters** [comment]
:+1: Looks good, besides a few minor remarks.

**Comment by Maxwe11** [comment]
@vweijsters updated. I'll squash feedback after yours "LGTM".

**Comment by vweijsters** [comment]
:+1: No further comments :smiley:

**Pull request merged**